subsys/logging/log_backend_adsp_mtrace.c

```diff
@@ -0,0 +1,252 @@
+/*
+ * Copyright (c) 2019,2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/logging/log_backend_adsp_mtrace.h>
+#include <zephyr/logging/log_backend.h>
+#include <zephyr/logging/log_core.h>
+#include <zephyr/logging/log_output.h>
+#include <zephyr/logging/log_backend_std.h>
+#include <zephyr/zephyr.h>
+#include <soc.h>
+#include <adsp_memory.h>
+
+/*
+ * A lock is needed as log_process() and log_panic() have no internal locks
+ * to prevent concurrency. Meaning if log_process is called after
+ * log_panic was called previously, log_process may happen from another
+ * CPU and calling context than the log processing thread running in
+ * the background. On an SMP system this is a race.
+ *
+ * This caused a race on the output trace such that the logging output
+ * was garbled and useless.
+ */
+static struct k_spinlock mtrace_lock;
+
+static uint32_t log_format_current = CONFIG_LOG_BACKEND_ADSP_MTRACE_OUTPUT_DEFAULT;
+
+static adsp_mtrace_log_hook_t mtrace_hook;
+
+static bool mtrace_active;
+
+/*
+ * SRAM window for mtrace is organized in 16 (equal sized) pages:
+ *
+ *	------------------------
+ *	| Page0  - descriptors |
+ *	------------------------
+ *	| Page1  - slot0       |
+ *	------------------------
+ *	| Page2  - slot1       |
+ *	------------------------
+ *	|         ...          |
+ *	------------------------
+ *	| Page14  - slot13     |
+ *	------------------------
+ *	| Page15  - slot14     |
+ *	------------------------
+ *
+ * The slot size == page size
+ *
+ * The first page contains descriptors for the remaining 15 cores
+ * The slot descriptor is:
+ * u32 res_id;
+ * u32 type;
+ * u32 vma;
+ *
+ * Log buffer slots have the following layout:
+ * u32 host_read_ptr;
+ * u32 dsp_write_ptr;
+ * u8 buffer[];
+ *
+ * The two pointers are offsets within the buffer.
+ */
+
+#define MTRACE_DESCRIPTOR_SIZE		12 /* 3 x u32 */
+#define MTRACE_PAGE_SIZE		0x1000
+#define MTRACE_SLOT_SIZE		MTRACE_PAGE_SIZE
+
+/* debug log slot types */
+#define MTRACE_SLOT_UNUSED			0x00000000
+#define MTRACE_SLOT_CRITICAL_LOG		0x54524300 /* byte 0: core ID */
+#define MTRACE_SLOT_DEBUG_LOG		0x474f4c00 /* byte 0: core ID */
+#define MTRACE_SLOT_GDB_STUB		0x42444700
+#define MTRACE_SLOT_TELEMETRY		0x4c455400
+#define MTRACE_SLOT_BROKEN			0x44414544
+ /* for debug and critical types */
+#define MTRACE_SLOT_CORE_MASK		GENMASK(7, 0)
+#define MTRACE_SLOT_TYPE_MASK		GENMASK(31, 8)
+
+#define MTRACE_LOG_BUF_SIZE		(MTRACE_SLOT_SIZE - 2 * sizeof(uint32_t))
+
+#define MTRACE_LOGGING_SLOT_TYPE(n)	(MTRACE_SLOT_DEBUG_LOG | (n & MTRACE_SLOT_CORE_MASK))
+
+#define MTRACE_DESCRIPTOR(n) ((uint8_t *)(z_soc_uncached_ptr((void *)HP_SRAM_WIN2_BASE) + n * MTRACE_DESCRIPTOR_SIZE))
```

Comment:
Performance: if you do something like `#define WIN2_UNCACHED_BASE z_soc_uncached_ptr((void *)HP_SRAM_WIN2_BASE)` then the offset gets done at compile time and not runtime, saving code size.  The dependency on "n" will defeat optimization unless "n" is known at build time.

Similar note more generically: everywhere this buffer is used, it's through the uncached pointers (unsurprisingly and correctly).  So IMHO it makes more sense just to construct them all from a single buffer base that is known to be uncached.

Reply:
Actually coming back, I don't see this macro actually used anywhere?  So maybe just remove?

Reply:
Ack, this was a code refactor leftover. I'll take note of the note on optimization though and check remain code.